# Lab 3
## Motion Capture Clustering

### Team members: Luay Dajani, Dana Geislinger, Chris Morgan, Caroll Rodriguez
##### Github - https://github.com/cdmorgan103/7331DataMiningNoShow

MSDS 7331, 12/02/2018

### Business Understanding 1
#### 10pts
Describe the purpose of the data set you selected (i.e., why was this data collected in the first place?). How will you measure the effectiveness of a good algorithm? Why does your chosen validation method make sense for this specific
dataset and the stakeholders needs?


https://archive.ics.uci.edu/ml/datasets/Motion+Capture+Hand+Postures#

A Vicon motion capture camera system was used to record 12 users performing 5 hand postures with markers attached to a left-handed glove. 

A rigid pattern of markers on the back of the glove was used to establish a local coordinate system for the hand, and 11 other markers were attached to the thumb and fingers of the glove. 3 markers were attached to the thumb with one above the thumbnail and the other two on the knuckles. 2 markers were attached to each finger with one above the fingernail and the other on the joint between the proximal and middle phalanx. 

The 11 markers not part of the rigid pattern were unlabeled; their positions were not explicitly tracked. Consequently, there is no a priori correspondence between the markers of two given records. In addition, due to the resolution of the capture volume and self-occlusion due to the orientation and configuration of the hand and fingers, many records have missing markers. Extraneous markers were also possible due to artifacts in the Vicon software's marker reconstruction/recording process and other objects in the capture volume. As a result, the number of visible markers in a record varied considerably. 

The data presented here is already partially preprocessed. First, all markers were transformed to the local coordinate system of the record containing them. Second, each transformed marker with a norm greater than 200 millimeters was pruned. Finally, any record that contained fewer than 3 markers was removed. The processed data has at most 12 markers per record and at least 3. For more information, see 'Attribute Information'. 

Due to the manner in which data was captured, it is likely that for a given record and user there exists a near duplicate record originating from the same user. We recommend therefore to evaluate classification algorithms on a leave-one-user-out basis wherein each user is iteratively left out from training and used as a test set. One then tests the generalization of the algorithm to new users. A 'User' attribute is provided to accomodate this strategy. 

This dataset may be used for a variety of tasks, the most obvious of which is posture recognition via classification. One may also attempt user identification. Alternatively, one may perform clustering (constrained or unconstrained) to discover marker distributions either as an attempt to predict marker identities or obtain statistical descriptions/visualizations of the postures. 

##### Attribute Information:

Data is provided as a CSV file. A header provides the name of each attribute. An initial dummy record composed entirely of 0s should be ignored. A question mark '?' is used to indicate a missing value. A record corresponds to a single instant or frame as recorded by the camera system. 

'Class' - Integer. The class ID of the given record. Ranges from 1 to 5 with 1=Fist(with thumb out), 2=Stop(hand flat), 3=Point1(point with pointer finger), 4=Point2(point with pointer and middle fingers), 5=Grab(fingers curled as if to grab). 
'User' - Integer. The ID of the user that contributed the record. No meaning other than as an identifier. 
'Xi' - Real. The x-coordinate of the i-th unlabeled marker position. 'i' ranges from 0 to 11. 
'Yi' - Real. The y-coordinate of the i-th unlabeled marker position. 'i' ranges from 0 to 11. 
'Zi' - Real. The z-coordinate of the i-th unlabeled marker position. 'i' ranges from 0 to 11. 

Each record is a set. The i-th marker of a given record does not necessarily correspond to the i-th marker of a different record. One may randomly permute the visible (i.e. not missing) markers of a given record without changing the set that the record represents. For the sake of convenience, all visible markers of a given record are given a lower index than any missing marker. A class is not guaranteed to have even a single record with all markers visible. 


### Data Understanding 1
#### 10pts
Describe the meaning and type of data (scale, values, etc.) for each attribute in the data file. Verify data quality: Are there missing values? Duplicate data? Outliers? Are those mistakes? How do you deal with these problems?

In [1]:
# Import required modules
import numpy as np
import pandas as pd
import math as math
import matplotlib.pyplot as plt
from matplotlib import style
from pprint import pprint
from IPython.display import display

# Load the data into variable 'df'
df = pd.read_csv('https://raw.githubusercontent.com/cdmorgan103/7331DataMiningNoShow/master/LAB3/Postures.csv')

# Get an overview of the raw data
df.info(null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 78096 entries, 0 to 78095
Data columns (total 38 columns):
Class    78096 non-null int64
User     78096 non-null int64
X0       78096 non-null float64
Y0       78096 non-null float64
Z0       78096 non-null float64
X1       78096 non-null float64
Y1       78096 non-null float64
Z1       78096 non-null float64
X2       78096 non-null float64
Y2       78096 non-null float64
Z2       78096 non-null float64
X3       78096 non-null object
Y3       78096 non-null object
Z3       78096 non-null object
X4       78096 non-null object
Y4       78096 non-null object
Z4       78096 non-null object
X5       78096 non-null object
Y5       78096 non-null object
Z5       78096 non-null object
X6       78096 non-null object
Y6       78096 non-null object
Z6       78096 non-null object
X7       78096 non-null object
Y7       78096 non-null object
Z7       78096 non-null object
X8       78096 non-null object
Y8       78096 non-null object
Z8       78096 n

In [2]:
# we will remove class 0 
df = df[df.Class !=0 ]

#change ? to none to improve data format
df=df.replace({'?': 'NaN'})

#coerces into numeric
df = df.apply(pd.to_numeric, errors='coerce')

In [3]:
df.head()

Class  User         X0         Y0         Z0         X1         Y1  \
1      1     0  54.263880  71.466776 -64.807709  76.895635  42.462500   
2      1     0  56.527558  72.266609 -61.935252  39.135978  82.538530   
3      1     0  55.849928  72.469064 -62.562788  37.988804  82.631347   
4      1     0  55.329647  71.707275 -63.688956  36.561863  81.868749   
5      1     0  55.142401  71.435607 -64.177303  36.175818  81.556874   

          Z1         X2         Y2 ...   Z8  X9  Y9  Z9  X10  Y10  Z10  X11  \
1 -72.780545  36.621229  81.680557 ...  NaN NaN NaN NaN  NaN  NaN  NaN  NaN   
2 -49.596509  79.223743  43.254091 ...  NaN NaN NaN NaN  NaN  NaN  NaN  NaN   
3 -50.606259  78.451526  43.567403 ...  NaN NaN NaN NaN  NaN  NaN  NaN  NaN   
4 -52.752784  86.320630  68.214645 ...  NaN NaN NaN NaN  NaN  NaN  NaN  NaN   
5 -53.475747  76.986143  42.426849 ...  NaN NaN NaN NaN  NaN  NaN  NaN  NaN   

   Y11  Z11  
1  NaN  NaN  
2  NaN  NaN  
3  NaN  NaN  
4  NaN  NaN  
5  NaN  NaN  

[5 rows x 38 columns]

In [4]:
# Get an overview of the raw data
df.info(null_counts=True)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 78095 entries, 1 to 78095
Data columns (total 38 columns):
Class    78095 non-null int64
User     78095 non-null int64
X0       78095 non-null float64
Y0       78095 non-null float64
Z0       78095 non-null float64
X1       78095 non-null float64
Y1       78095 non-null float64
Z1       78095 non-null float64
X2       78095 non-null float64
Y2       78095 non-null float64
Z2       78095 non-null float64
X3       77405 non-null float64
Y3       77405 non-null float64
Z3       77405 non-null float64
X4       74975 non-null float64
Y4       74975 non-null float64
Z4       74975 non-null float64
X5       65072 non-null float64
Y5       65072 non-null float64
Z5       65072 non-null float64
X6       52247 non-null float64
Y6       52247 non-null float64
Z6       52247 non-null float64
X7       38943 non-null float64
Y7       38943 non-null float64
Z7       38943 non-null float64
X8       30563 non-null float64
Y8       30563 non-null float64

Only 31 data points appear for the 11th sensor, and are for one user. We will delete this as a feature as it does not provide any value.

In [5]:
del df['X11']
del df['Y11']
del df['Z11']

   begin work on fix

In [6]:
pd.set_option("display.max_columns",100)

In [7]:
df.head()

Class  User         X0         Y0         Z0         X1         Y1  \
1      1     0  54.263880  71.466776 -64.807709  76.895635  42.462500   
2      1     0  56.527558  72.266609 -61.935252  39.135978  82.538530   
3      1     0  55.849928  72.469064 -62.562788  37.988804  82.631347   
4      1     0  55.329647  71.707275 -63.688956  36.561863  81.868749   
5      1     0  55.142401  71.435607 -64.177303  36.175818  81.556874   

          Z1         X2         Y2         Z2         X3         Y3  \
1 -72.780545  36.621229  81.680557 -52.919272  85.232264  67.749220   
2 -49.596509  79.223743  43.254091 -69.982489  87.450873  68.400808   
3 -50.606259  78.451526  43.567403 -70.658489  86.835388  68.907925   
4 -52.752784  86.320630  68.214645 -72.228461  61.596157  11.250648   
5 -53.475747  76.986143  42.426849 -72.574743  86.368748  67.901260   

          Z3         X4         Y4         Z4  X5  Y5  Z5  X6  Y6  Z6  X7  Y7  \
1 -73.684130  59.188576  10.678936 -71.297781 NaN NaN NaN NaN NaN NaN NaN NaN   
2 -70.703991  61.587452  11.779919 -68.827418 NaN NaN NaN NaN NaN NaN NaN NaN   
3 -71.138344  61.686427  11.793440 -68.889316 NaN NaN NaN NaN NaN NaN NaN NaN   
4 -68.956425  77.387225  42.717833 -72.015146 NaN NaN NaN NaN NaN NaN NaN NaN   
5 -72.444650  61.275402  10.841109 -69.279906 NaN NaN NaN NaN NaN NaN NaN NaN   

   Z7  X8  Y8  Z8  X9  Y9  Z9  X10  Y10  Z10  
1 NaN NaN NaN NaN NaN NaN NaN  NaN  NaN  NaN  
2 NaN NaN NaN NaN NaN NaN NaN  NaN  NaN  NaN  
3 NaN NaN NaN NaN NaN NaN NaN  NaN  NaN  NaN  
4 NaN NaN NaN NaN NaN NaN NaN  NaN  NaN  NaN  
5 NaN NaN NaN NaN NaN NaN NaN  NaN  NaN  NaN

In [8]:
df.describe()

Class          User            X0            Y0            Z0  \
count  78095.000000  78095.000000  78095.000000  78095.000000  78095.000000   
mean       2.983776      7.959229     50.346308     85.813150    -29.985096   
std        1.421152      4.697754     32.695886     40.203448     34.361971   
min        1.000000      0.000000   -108.552738    -98.233756   -126.770872   
25%        2.000000      5.000000     29.295141     63.497746    -56.356593   
50%        3.000000      9.000000     54.620245     86.526334    -30.864248   
75%        4.000000     12.000000     72.488833    113.108673     -1.419462   
max        5.000000     14.000000    190.017835    169.175464    113.345119   

                 X1            Y1            Z1            X2            Y2  \
count  78095.000000  78095.000000  78095.000000  78095.000000  78095.000000   
mean      49.595844     86.193751    -29.509579     48.612744     83.772387   
std       32.477961     40.452297     34.764460     33.605155     41.022710   
min     -111.685241    -96.142589   -166.006838   -106.886524   -100.789312   
25%       28.755679     64.156450    -57.360408     25.173405     58.053733   
50%       54.215714     87.543493    -30.185331     53.814592     86.459935   
75%       71.763080    116.228881     -0.368080     71.561988    106.661720   
max      188.691997    170.209350    104.697852    188.760168    168.186466   

                 Z2            X3            Y3            Z3            X4  \
count  78095.000000  77405.000000  77405.000000  77405.000000  74975.000000   
mean     -30.560906     48.492574     82.064324    -31.146643     48.406087   
std       35.120384     33.873116     41.571227     35.933363     34.226361   
min     -129.595296   -111.761053    -97.603414   -143.540529    -99.107635   
25%      -58.654339     23.861711     54.072442    -59.369655     22.829056   
50%      -32.356535     54.136687     85.749842    -34.015893     53.856778   
75%       -0.946134     71.435593    105.531551     -1.439540     71.970195   
max      104.590879    151.033472    168.292018    129.316870    172.275978   

                 Y4            Z4            X5            Y5            Z5  \
count  74975.000000  74975.000000  65072.000000  65072.000000  65072.000000   
mean      80.419506    -31.993556     47.042778     81.394937    -30.270370   
std       42.623146     36.398277     34.815074     42.864544     36.853118   
min      -97.948829   -157.199089   -120.657868    -97.468548   -135.699430   
25%       48.960422    -60.548201     19.538848     49.252466    -58.994392   
50%       85.777670    -35.165459     51.951879     88.137089    -31.714450   
75%      105.483241     -2.141662     71.888850    107.835725      0.117398   
max      168.258643    119.237203    180.563322    167.926171    110.898899   

                 X6            Y6            Z6            X7            Y7  \
count  52247.000000  52247.000000  52247.000000  38943.000000  38943.000000   
mean      45.677699     83.732313    -26.639238     44.463433     88.463343   
std       36.342190     43.129421     35.709681     38.378395     40.503378   
min     -100.084275    -67.283707   -153.449813   -108.605639    -64.972157   
25%       15.486358     51.664992    -55.374994     13.081363     63.541846   
50%       52.085617     92.154280    -26.457327     49.601058     93.796334   
75%       72.615455    111.473915      2.433526     75.721811    119.188963   
max      176.409004    168.598384    117.914907    189.221529    169.127359   

                 Z7            X8            Y8            Z8            X9  \
count  38943.000000  30563.000000  30563.000000  30563.000000  23967.000000   
mean     -20.368719     48.209394     86.030582    -24.364841     54.749001   
std       33.943957     38.604623     41.613479     35.736660     39.969941   
min     -113.733105   -121.182089    -65.077550   -142.654497    -99.231688   
25%      -45.676522     17.044171     53.669158    -52.5217

In [9]:
xcols= ('X0','X1','X2','X3','X4','X5','X6','X7','X8','X9','X10')
ycols= ('Y0','Y1','Y2','Y3','Y4','Y5','Y6','Y7','Y8','Y9','Y10')
zcols= ('Z0','Z1','Z2','Z3','Z4','Z5','Z6','Z7','Z8','Z9','Z10')
dfx=df[['X0','X1','X2','X3','X4','X5','X6','X7','X8','X9','X10']].copy()
dfy=df[['Y0','Y1','Y2','Y3','Y4','Y5','Y6','Y7','Y8','Y9','Y10']].copy()
dfz=df[['Z0','Z1','Z2','Z3','Z4','Z5','Z6','Z7','Z8','Z9','Z10']].copy()

In [10]:
dfx['xmin']=dfx.min(axis=1)
dfy['ymin']=dfy.min(axis=1)
dfz['zmin']=dfz.min(axis=1)

In [11]:
dfx.head()

X0         X1         X2         X3         X4  X5  X6  X7  X8  X9  \
1  54.263880  76.895635  36.621229  85.232264  59.188576 NaN NaN NaN NaN NaN   
2  56.527558  39.135978  79.223743  87.450873  61.587452 NaN NaN NaN NaN NaN   
3  55.849928  37.988804  78.451526  86.835388  61.686427 NaN NaN NaN NaN NaN   
4  55.329647  36.561863  86.320630  61.596157  77.387225 NaN NaN NaN NaN NaN   
5  55.142401  36.175818  76.986143  86.368748  61.275402 NaN NaN NaN NaN NaN   

   X10       xmin  
1  NaN  36.621229  
2  NaN  39.135978  
3  NaN  37.988804  
4  NaN  36.561863  
5  NaN  36.175818

In [ ]:
dfxxmin=dfx.min(axis=1)
dfyymin=dfy.min(axis=1)
dfzzmin=dfz.min(axis=1)

In [ ]:
dfx=dfx.values
dfxxmin=dfxxmin.values

In [ ]:
type(dfx)
print(dfx)
dfx.shape

In [ ]:
dfx['lowest_colx'] = dfx.idxmin(axis=1)
dfy['lowest_coly'] = dfy.idxmin(axis=1)
dfz['lowest_colz'] = dfz.idxmin(axis=1)

In [ ]:

for i in dfx.columns:
    dfx[a,i] =dfx[a,i] - dfx.xmin[a,i]

In [ ]:
dfx.columns

In [ ]:
###x values
dfx.X0=dfx.X0-dfx.xmin
dfx.X1=dfx.X1-dfx.xmin
dfx.X2=dfx.X2-dfx.xmin
dfx.X3=dfx.X3-dfx.xmin
dfx.X4=dfx.X4-dfx.xmin
dfx.X5=dfx.X5-dfx.xmin
dfx.X6=dfx.X6-dfx.xmin
dfx.X7=dfx.X7-dfx.xmin
dfx.X8=dfx.X8-dfx.xmin
dfx.X9=dfx.X9-dfx.xmin
dfx.X10=dfx.X10-dfx.xmin

###y values
dfy.Y0=dfy.Y0-dfy.ymin
dfy.Y1=dfy.Y1-dfy.ymin
dfy.Y2=dfy.Y2-dfy.ymin
dfy.Y3=dfy.Y3-dfy.ymin
dfy.Y4=dfy.Y4-dfy.ymin
dfy.Y5=dfy.Y5-dfy.ymin
dfy.Y6=dfy.Y6-dfy.ymin
dfy.Y7=dfy.Y7-dfy.ymin
dfy.Y8=dfy.Y8-dfy.ymin
dfy.Y9=dfy.Y9-dfy.ymin
dfy.Y10=dfy.Y10-dfy.ymin

###y values
dfz.Z0=dfz.Z0-dfz.zmin
dfz.Z1=dfz.Z1-dfz.zmin
dfz.Z2=dfz.Z2-dfz.zmin
dfz.Z3=dfz.Z3-dfz.zmin
dfz.Z4=dfz.Z4-dfz.zmin
dfz.Z5=dfz.Z5-dfz.zmin
dfz.Z6=dfz.Z6-dfz.zmin
dfz.Z7=dfz.Z7-dfz.zmin
dfz.Z8=dfz.Z8-dfz.zmin
dfz.Z9=dfz.Z9-dfz.zmin
dfz.Z10=dfz.Z10-dfz.zmin


In [ ]:
dfx.tail()

In [ ]:
dfxyz=df.copy()

In [ ]:
dfxyz

In [ ]:
dfz.tail()

In [ ]:
dfx=df.copy()
del df['X11']
del df['Y11']
del df['Z11']

In [ ]:
####alt version

In [ ]:
df['tempx'] = 0-df.X0
df['tempy'] = 0-df.Y0
df['tempz'] = 0-df.Z0

In [ ]:
dfcen=df.copy()

In [ ]:
###x values
dfcen.X0=dfcen.X0+df.tempx
dfcen.X1=dfcen.X1+df.tempx
dfcen.X2=dfcen.X2+df.tempx
dfcen.X3=dfcen.X3+df.tempx
dfcen.X4=dfcen.X4+df.tempx
dfcen.X5=dfcen.X5+df.tempx
dfcen.X6=dfcen.X6+df.tempx
dfcen.X7=dfcen.X7+df.tempx
dfcen.X8=dfcen.X8+df.tempx
dfcen.X9=dfcen.X9+df.tempx
dfcen.X10=dfcen.X10+df.tempx

###y values
dfcen.Y0=dfcen.Y0+df.tempy
dfcen.Y1=dfcen.Y1+df.tempy
dfcen.Y2=dfcen.Y2+df.tempy
dfcen.Y3=dfcen.Y3+df.tempy
dfcen.Y4=dfcen.Y4+df.tempy
dfcen.Y5=dfcen.Y5+df.tempy
dfcen.Y6=dfcen.Y6+df.tempy
dfcen.Y7=dfcen.Y7+df.tempy
dfcen.Y8=dfcen.Y8+df.tempy
dfcen.Y9=dfcen.Y9+df.tempy
dfcen.Y10=dfcen.Y10+df.tempy

###y values
dfcen.Z0=dfcen.Z0+df.tempz
dfcen.Z1=dfcen.Z1+df.tempz
dfcen.Z2=dfcen.Z2+df.tempz
dfcen.Z3=dfcen.Z3+df.tempz
dfcen.Z4=dfcen.Z4+df.tempz
dfcen.Z5=dfcen.Z5+df.tempz
dfcen.Z6=dfcen.Z6+df.tempz
dfcen.Z7=dfcen.Z7+df.tempz
dfcen.Z8=dfcen.Z8+df.tempz
dfcen.Z9=dfcen.Z9+df.tempz
dfcen.Z10=dfcen.Z10+df.tempz


In [ ]:
dfcen.head()

In [ ]:
dfcen.describe()

In [ ]:
dfcen

#### Reformatting the Data for Sensor Clustering

The raw data for this dataset is formatted with each capture being treated as a separate observation. This makes sense if you are classifying the gesture or user, but we are trying to cluster in order to identify which sensor on the glove corresponds to a specific set of coordinates. Therefore, we must stack the data in order to get a new dataset where each observation has only a single X, Y, and Z coordinate, and corresponds to the specific position of an individual sensor module in any given capture. 

The original dataset consists of 78,095 captures with coordinates on up to 11 sensors in any given capture, as well as metadata pertaining to the capture about the user and class the capture was representative of. We will retain this metadata, and also create a new variable that corresponds to the sensor number (0 through 10). However, each coordinate will now be labelled as the raw coordinate number on each cartesian axis (X, Y, Z) regardless of sensor number (as opposed to X0, Y10, Z5, etc. as was the case originally).

At this point, we will now also drop missing datapoints; NaN values in this dataset represent sensors that were obscured or not recorded for some other reason during a capture, and since we are interested in clustering by sensor, it does not further the analysis to include sensor coordinates for which no data was recorded.

In [ ]:
# Create a list of dataframes with X, Y, Z coordinates selected for each individual sensor (0-10)
df_coords = [df.loc[:, ['Class', 'User', 'X%d' % i, 'Y%d' %i, 'Z%d' % i]] for i in range(11)]

for i in range(len(df_coords)):
    # NaN values are dropped at this time
    df_coords[i] = df_coords[i].dropna()
    
    # Coordinate variable labels will now be standardized
    df_coords[i].columns = ['Class', 'User', 'X', 'Y', 'Z']
    
    # Sensor number will now be added as a variable 'sensor' to each dataset
    df_coords[i]['Sensor'] = i
    
# Combine datasets into new dataframe: df2
df2 = pd.concat(df_coords, ignore_index=True)

# Reorder columns with metadata columns first
df2 = df2[['Class', 'User', 'Sensor', 'X', 'Y', 'Z']]

With the newly formatted dataset, we now have 612,209 distinct cartesian coordinates with corresponding Sensor, Class, and User metadata.

In [ ]:
df2.info()

### Data Understanding 2
#### 10pts
Visualize the any important attributes appropriately. Important: Provide an interpretation for any charts or graphs.

##### Number of Datapoints per User by Class (Hand Gesture)

In [ ]:
import pandas as pd
# Print descriptive info for the unique values for each predictor

#1=Fist(with thumb out), 
#2=Stop(hand flat), 
#3=Point1(point with pointer finger), 
#4=Point2(point with pointer and middle fingers), 
#5=Grab(fingers curled as if to grab). 

print('Class:', list(df2.Class.unique()))
print('User:', list(df2.User.unique()))
#df_User = df.groupby(['User','Class'])['Class'].count() 
df_Class = df2.groupby(['Class', 'User'])['User'].count()

df_Class['total'] = sum(df_Class) #df_User 

df_Class

In [ ]:
del df_Class['total']
#plot and format bar chart
plt.figure(figsize=(16,5), dpi=80)
_ = df_Class.plot(kind='bar')
#plt.barh(df_neighborhood, width=0.4, height=0.4, align='center', alpha=0.4)

### Percentage of datapoints for each User by Class (Hand Gesture)

In [ ]:
byClass = df2.groupby(['Class'])['User'].count()
User_byClass = df2.groupby(['Class','User'])['User'].count()

output = User_byClass['Percentage'] = (User_byClass / byClass ) * 100

df_output = output.to_frame("Percentage")
df_output

### Interactive Plotting of User/Class Combinations

In [ ]:
import plotly
import plotly.plotly as py
import plotly.graph_objs as go
from plotly.offline import init_notebook_mode
import pandas as pd
import numpy as np

# Set up plotly API key
plotly.tools.set_credentials_file(username='danageis', api_key='0NAcrzM8YksRAT3uyEZ7')

# Enable offline plots (embed into ipynb)
init_notebook_mode(connected=True)

# Define list of possible users
possible_users = set(df.User.unique())

# Define dictionary of all possible user/class combinations (which ones took pictures of which)
possible_classes = {usr: set(df.where(df.User == usr).Class.dropna().unique()) for usr in possible_users}

# Define functions to create 3D-interactive plots of each user/position combo
def parse_coords(cls=1, usr=1):
    """ Helper function: return all coords for specified User and Class
    
    cls (int): Class (defaults to 1)
    usr (int): User (defaults to 1)
    
    Ret (dict): {dimension (str): coordinates (pd.Series)}
    """
    dims = ['X', 'Y', 'Z']
    fill = lambda suffix: [suffix + str(i) for i in range(1,11)]
    cols = {col: fill(col) for col in dims}
    coords = {dim: df[cols[dim]].where(df.User == usr).where(df.Class == cls).stack() for dim in dims}
    return coords

def plot_3d(cls=1, usr=1, title=None, allpts=False, clusters=np.array([0])):
    """ Return 3D interactive plot for given coordinates set (single user/sign combos)
    
    cls (int): Class (defaults to 1)
    usr (int): User (defaults to 1)
    title (str): Title of the plot (defaults to "User {usr} Position {cls}")
    allpts (bool): Whether or not this is meant to cluster all points in the dataset (df2)
        All 600k+ points is too many to plot interactively; this will sample 20,000 points randomly and graph them
    clusters (np.array): Clusters to use for color-coding (only valid for allpts at this time)
    
    Ret (plotly.graph_objs.Scatter3d figure object)
    """
    
    # Make sure user/class designation is valid
    if not usr in possible_users:
        return "ERROR: User %d does not have any data. Try again with a different user specified." % usr
    elif not cls in possible_classes[usr]:
        return "ERROR: User %d does not have any data for Class %d. Try again with a different class specified." % (usr, cls)
    
    # Process points to draw
    if allpts:
        # Sample 20,000 random coords w/o replacement
        coords = df2.copy()
        if clusters.any():
            coords['cluster'] = clusters
        coords = coords.sample(n=20000, replace=False)
    else:
        coords = parse_coords(cls, usr)
        
    # Name plot if title not provided (default behavior)
    if not title:
        if allpts:
            title = "20,000 Random Points from Dataset"
            if clusters.any():
                title += " (Clustered)"
        else:
            title = "User %d Position %d" % (usr, cls)

    # Color-code clustered points, if provided
    if allpts and clusters.any():
        marker_params = dict(size=5,
                             color=coords.cluster,
                             colorscale='Portland'
                             )
    else:
        marker_params = dict(size=5)
            
    data = [go.Scatter3d(x=coords['X'],
                         y=coords['Y'],
                         z=coords['Z'],
                         mode='markers',
                         marker=marker_params
                         )
            ]

    layout = dict(width=800,
                  height=700,
                  autosize=True,
                  title=title,
                  scene=dict(xaxis=dict(gridcolor='rgb(255, 255, 255)',
                                        zerolinecolor='rgb(255, 255, 255)',
                                        showbackground=True,
                                        backgroundcolor='rgb(230, 230,230)'
                                        ),
                             yaxis=dict(gridcolor='rgb(255, 255, 255)',
                                        zerolinecolor='rgb(255, 255, 255)',
                                        showbackground=True,
                                        backgroundcolor='rgb(230, 230,230)'
                                        ),
                             zaxis=dict(gridcolor='rgb(255, 255, 255)',
                                        zerolinecolor='rgb(255, 255, 255)',
                                        showbackground=True,
                                        backgroundcolor='rgb(230, 230,230)'
                                        ),
                             camera=dict(up=dict(x=0,
                                                 y=0,
                                                 z=1
                                                 ),
                                         eye=dict(x=-1.7428,
                                                  y=1.0707,
                                                  z=0.7100,
                                                  )
                                         ),
                             aspectratio = dict(x=1,
                                                y=1,
                                                z=0.7
                                                ),
                             aspectmode = 'manual'
                             ),
                )

    fig = dict(data=data, layout=layout)

    return py.iplot(fig, filename="Hands-%s" % title)

In [ ]:
from warnings import filterwarnings
filterwarnings('ignore')

In [ ]:
plot_3d(cls=1, usr=1)

### Modeling and Evaluation 1
#### 10 pts
Train and adjust parameters

For this data, we will be attempting to cluster the data on the basis of the sensors used to collect the data. As described in the business understanding section, this dataset consists of snapshot measurements taken from a special glove fitted with sensors to track a person's hand movements and map them to 3-dimensional space. The dataset has recorded data for up to 11 sensors in each capture (some may have been obscured when measurement was performed), and associated 3D coordinates for each sensor. Captures were taken for 13 users who were performing one of five pre-defined gestures with their hand.

Our goal is to cluster each data point based on which of the 11 possible sensors it corresponds to. Each record in the dataset has been formatted so that it represents a single coordinate location for an individual sensor. To accomplish this we will attempt clustering with 11 total clusters defined in our clustering algorithms, ideally corresponding with each of the sensors that is present on the glove. We will begin by performing cluster analysis of all points aggregated together:

In [ ]:
plot_3d(allpts=True)

In [ ]:
plot_3d(allpts=True)

Shown above are 2 random samplings of 20,000 points each from the dataset. This was done to illustrate that when multiple random samples are plotted in 3-dimensional space, the distributions are different each time but there are similar patterns that can be observed consistently such as 2 large elliptoid clusters on the right side and some smaller clusters on the left side. This provides some anecdotal evidence that clustering may be successful using all available data aggregated from every user and every gesture class.

### GMM Clustering

Visualizing the data set as we have in the two random samples above, it seems that while there are some obvious discrete clusters in 3-dimensional space these clusters are not necessarily uniform globular clusters. These type of clusters might be well defined using Gaussian mixture model clustering which would allow for normally distributed clusters of varying shapes and sizes. We will first attempt to fit a Gaussian mixture model to obtain clusters for the 11 sensors in the dataset.

In [ ]:
from sklearn.mixture import GaussianMixture as GMM

# Fit 11 clusters
#  Set random state to '123456789' to make sure same result achieved each time
gmm = GMM(n_components=11,
          random_state=123456789
          )
gmm_X = df2[['X','Y','Z']]
gmm.fit(gmm_X)
gmm_clusters = gmm.predict(gmm_X)

In [ ]:
from mpl_toolkits import mplot3d
%matplotlib inline
import numpy as np
import matplotlib.pyplot as plt
ax = plt.axes(projection='3d')

# Data for a three-dimensional line
zline = np.linspace(0, 15, 1000)
xline = np.sin(zline)
yline = np.cos(zline)
ax.plot3D(xline, yline, zline, 'gray')

# Data for three-dimensional scattered points
zdata = gmm_X['Z'].astype(float)
xdata = gmm_X['X'].astype(float)
ydata = gmm_X['Y'].astype(float)
_ = ax.scatter3D(gmm_X['X'], gmm_X['Y'], gmm_X['Z'], c=gmm_clusters)

In [ ]:
plot_3d(allpts=True, clusters=gmm_clusters)

We plot the basic GMM clusters above, both as a static image consisting of all points (more than 600,000) and an interactive plot that is sampled as 20,000 random points from the dataset color-coded by cluster. From these visuals, we can see that there does appear to be decent clustering.

From the data description, we know that there are 2 sensors on each finger and 3 sensors on the thumb. We also know that there are several positions in which multiple (sometimes more than half) of the sensors are obscured from view and not recorded. Therefore, this clustering makes sense from a logical perspective, but it could likely still be improved.

Another interesting comparison is the distribution of clusters as defined by GMM. While we do not know from the raw data which sensors correspond to which point on the glove, we do know that in a majority of recorded instances there are sensors that were unable to be recorded. It would be interesting to note if this clustering algorithm created clusters of varying sizes, which might be indicative of an uneven distribution of which sensors were not recorded most or least often.

In [ ]:
# Create a new dataframe with the clusters defined as a new variable
df_gmm = gmm_X
df_gmm['Clusters'] = gmm_clusters

# Count the number of instances in each cluster and convert to a percentage of the total
cluster_pcts = df_gmm.groupby('Clusters').count()['X'] / len(gmm_X) * 100

# Plot the distribution
ax = cluster_pcts.plot(kind='bar', title="Percentage of Each GMM-Defined Cluster")
_ = ax.set_ylabel("Percentage of Total Points (%)")

It is interesting to note from this distribution that there are significantly varying distributions between clusters. If these clusters do in fact coincide with sensors on the glove, this could provide evidence that certain sensors appear more often than others.

This also makes logical sense, since it would be expected that some of the gestures being performed would have obscured at least part of the hand. The distribution shows that from GMM, there are 2 clusters defined that are very scarce in the total (both less than 5%) compared to the rest. This could be accounted for as the 2 sensors present on a finger that is often obscured, such as the pinky finger. Conversely, there are 2 clusters that stand out as about the same percentage as eachother (almost 14%) but both significantly higher than most of the other clusters. These cluster could represent sensors on fingers that are needed to perform most of the gestures recorded, so they might show up regardless of user or class.

### Modeling and Evaluation 2
#### 10pts
Evaluate and Compare

#####


In [ ]:
from sklearn import metrics
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import AgglomerativeClustering
from sklearn.cluster import KMeans
from sklearn.cluster import DBSCAN


data = np.array(df2.to_array())

n_clusters = 11

model = KMeans(init='k-means++', n_clusters=n_clusters, n_init=1).fit(data)
# model = AgglomerativeClustering(linkage='ward', n_clusters=n_clusters).fit(data)
model = DBSCAN(eps=0.15, min_samples=10).fit(data)

labels = model.labels_
n_clusters_ = len(set(labels)) - (1 if -1 in labels else 0)

print('Estimated number of clusters: %d' % n_clusters_)
print("Homogeneity: %0.3f" % metrics.homogeneity_score(labels_true, labels))
print("Completeness: %0.3f" % metrics.completeness_score(labels_true, labels))
print("V-measure: %0.3f" % metrics.v_measure_score(labels_true, labels))
print("Adjusted Rand Index: %0.3f"
      % metrics.adjusted_rand_score(labels_true, labels))
print("Adjusted Mutual Information: %0.3f"
      % metrics.adjusted_mutual_info_score(labels_true, labels))
print("Silhouette Coefficient: %0.3f"
      % metrics.silhouette_score(data, labels))

silhouette_per_sample = metrics.silhouette_samples(data,labels)

# get the middle of each cluster
centroids = []
for lab in range(0,n_clusters_):
    centroids.append( [np.mean(data[np.where(labels==lab),0]), np.mean(data[np.where(labels==lab),1])] )
centroids = np.array(centroids)

plt.figure()
plt.scatter(data[:, 0], data[:, 1], c=labels,
                    cmap=plt.cm.spectral, s=5, linewidths=0)

if len(centroids)>0:
    plt.scatter(centroids[:, 0], centroids[:, 1],
                marker='+', s=200, linewidths=3, color='k', zorder=10)  # plot the centroids
plt.title('Estimated number of clusters: %d' % n_clusters_)

plt.figure()
plt.scatter(data[:, 0], data[:, 1], c=silhouette_per_sample,
                cmap=plt.cm.gray, s=5, linewidths=0)
plt.title('Clusters with silhouette coefficient coloring')
plt.show()

In [ ]:
# do it for the k-means
from sklearn.cluster import KMeans
data = df2
seuclid = []
scosine = []
k = range(2,10)
for i in k:
    kmeans_model = KMeans(n_clusters=i, init="k-means++").fit(data)
    labels = kmeans_model.labels_
    seuclid.append(metrics.silhouette_score(data, labels, metric='euclidean'))
    scosine.append(metrics.silhouette_score(data, labels, metric='cosine'))
    
plt.figure(figsize=(10,5))
plt.plot(k,seuclid,label='euclidean')
plt.plot(k,scosine,label='cosine')
plt.ylabel("Silhouette")
plt.xlabel("Cluster")
plt.title("Silhouette vs Cluster Size")
plt.legend()
plt.show()

### Modeling and Evaluation 3
#### 10 pts
Visualize Results

### Modeling and Evaluation 4
#### 20 pts
Summarize the Ramifications

### Deployment 
#### 10 pts
Be critical of your performance and tell the reader how you current model might be usable by other parties. Did you achieve your goals? If not, can you reign in the utility of your modeling? How useful is your model for interested parties (i.e., the companies or organizations that might want to use it)? How would your deploy your model for interested parties? What other data should be collected? How often would the model need to be updated, etc.?

### Exceptional Work
#### 10 pts
You have free reign to provide additional analyses or combine analyses.

In [ ]:
from mpl_toolkits import mplot3d

In [ ]:
%matplotlib inline
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
#1=Fist(with thumb out), 
#2=Stop(hand flat), 
#3=Point1(point with pointer finger), 
#4=Point2(point with pointer and middle fingers), 
#5=Grab(fingers curled as if to grab). 

#filter certain users for plotting below
df_user6 = df[df.User == 6][df.Class == 2]
df_user14 = df[df.User == 14][df.Class == 1]
df_user5 = df[df.User == 5][df.Class == 4]
df_user8 = df[df.User == 8][df.Class == 3]

#One row in data table - represnts on set of data points
df_row48977 = df[48976:48977][df.Class == 5]

# feature categories
X_features = ['X0','X1','X2','X3','X4','X5','X6','X7','X8','X9','X10']
Y_features = ['Y0','Y1','Y2','Y3','Y4','Y5','Y6','Y7','Y8','Y9','Y10']
Z_features = ['Z0','Z1','Z2','Z3','Z4','Z5','Z6','Z7','Z8','Z9','Z10']

##### One record plotted in 3D show Class 5 (Grab Gestures)

In [ ]:
#one set of markers
ax = plt.axes(projection='3d')

# Data for a three-dimensional line
zline = np.linspace(0, 15, 1000)
xline = np.sin(zline)
yline = np.cos(zline)
ax.plot3D(xline, yline, zline, 'gray')

# Data for three-dimensional scattered points
zdata = df_row48977[Z_features].astype(float)
xdata = df_row48977[X_features].astype(float)
ydata = df_row48977[Y_features].astype(float)
ax.scatter3D(xdata, ydata, zdata);

##### All User 6's records plotted in 3D with Posture 3 (Stop sign)

In [ ]:
ax = plt.axes(projection='3d')

# Data for a three-dimensional line
zline = np.linspace(0, 15, 1000)
xline = np.sin(zline)
yline = np.cos(zline)
ax.plot3D(xline, yline, zline, 'gray')

# Data for three-dimensional scattered points
zdata = df_user6[Z_features].astype(float)
xdata = df_user6[X_features].astype(float)
ydata = df_user6[Y_features].astype(float)
ax.scatter3D(xdata, ydata, zdata);

##### All User 14's records plotted in 3D with Class 1 (Fist Gesture)

In [ ]:
ax = plt.axes(projection='3d')

# Data for a three-dimensional line
zline = np.linspace(0, 15, 1000)
xline = np.sin(zline)
yline = np.cos(zline)
ax.plot3D(xline, yline, zline, 'gray')

# Data for three-dimensional scattered points
zdata = df_user14[Z_features].astype(float)
xdata = df_user14[X_features].astype(float)
ydata = df_user14[Y_features].astype(float)
ax.scatter3D(xdata, ydata, zdata);

##### All User 8's records plotted in 3D with Class 3 (point with pointer finger Gesture)

In [ ]:
ax = plt.axes(projection='3d')

# Data for a three-dimensional line
zline = np.linspace(0, 15, 1000)
xline = np.sin(zline)
yline = np.cos(zline)
ax.plot3D(xline, yline, zline, 'gray')

# Data for three-dimensional scattered points
zdata = df_user8[Z_features].astype(float)
xdata = df_user8[X_features].astype(float)
ydata = df_user8[Y_features].astype(float)
ax.scatter3D(xdata, ydata, zdata);

##### All User 5's records plotted in 3D with Class 5 (Grab Gesture)

In [ ]:
ax = plt.axes(projection='3d')

# Data for a three-dimensional line
zline = np.linspace(0, 15, 1000)
xline = np.sin(zline)
yline = np.cos(zline)
ax.plot3D(xline, yline, zline, 'gray')

# Data for three-dimensional scattered points
zdata = df_user5[Z_features].astype(float)
xdata = df_user5[X_features].astype(float)
ydata = df_user5[Y_features].astype(float)
ax.scatter3D(xdata, ydata, zdata);

##### All User 5's records plotted in 3D in a Wireframe show Class 2 (Stop Gesture)

In [ ]:
fig = plt.figure()
ax = plt.axes(projection='3d')
ax.plot_wireframe(df_user6[X_features], df_user6[Y_features], df_user6[Z_features], color='black')
ax.set_title('wireframe');
